docs/source/worker-state.rst

```diff
@@ -0,0 +1,433 @@
+task into a thread from a local thread pool to execute.
+
+Optionally, while it's running, this task may identify itself as a long-running task
+(see :doc:`Tasks launching tasks <task-launch>`), at which point it secedes from the
+thread pool and changes state to `long-running`. ``executing`` and ``long-running`` are
+almost identical states, the only difference being that the latter don't count towards
+the maximum number of tasks running in parallel at the same time.
+
+A task can terminate in three ways:
+
+- Complete successfully; its return value is stored in :attr:`~WorkerState.data`
+- Raise an exception
+- Raise :class:`~distributed.Reschedule`; it is immediately forgotten.
+
+In all cases, the outcome is sent back to the scheduler.
+
+
+Scattered data
+--------------
+:meth:`Scattered data <distributed.Client.scatter>` follows an even simpler path,
+landing directly in ``memory``:
+
+.. image:: images/worker-scatter-state.svg
+    :alt: Worker states for scattered data
+
+
+Forgetting tasks
+----------------
+Once a task is in ``memory`` or ``error``, the Worker will hold onto it indefinitely,
+until the Scheduler explicitly asks the Worker to forget it.
+This happens when there are no more Clients holding a reference to the key and there are
+no more waiter tasks (that is, dependents that have not been computed). Additionally,
+the :doc:`active_memory_manager` may ask to drop excess replicas of a task.
+
+
+.. image:: images/worker-forget-state.svg
+    :alt: Worker states for computing tasks
+
+
+Irregular flow
+--------------
+There are a few important exceptions to the flow diagrams above:
+
+- A task is `stolen <work-stealing>`_, in which case it transitions from ``waiting``,
+  ``ready``, or ``constrained`` directly to ``released``. Note that steal requests for
+  tasks that are currently executing are rejected.
+- Scheduler intercession, in which the scheduler reassigns a task that was
+  previously assigned to a separate worker to a new worker. This most commonly
+  occurs when a :doc:`worker dies <killed>` during computation.
+- Client intercession, where a client either explicitly releases a Future or descopes
+  it; alternatively the whole client may shut down or become unresponsive. When there
+  are no more clients holding references to a key or one of its dependents, the
+  Scheduler will release it.
+
+In short:
+
+.. important::
+   A task can transition to ``released`` from *any* state, not just those in the
+   diagrams above.
+
+If there are no dependants, the task immediately transitions to ``forgotten`` and is
+descoped. However, there is an important exception, :ref:`cancelled-tasks`.
+
+
+.. _cancelled-tasks:
+
+Task cancellation
+-----------------
+The Worker may receive a request to release a key while it is currently in ``flight``,
+``executing``, or ``long-running``. Due to technical limitations around cancelling
+Python threads, such an event cannot cause the related asyncio task to be immediately
```

Comment:
```suggestion
Python threads, and the way data fetching from peer workers is currently implemented, such an event cannot cause the related asyncio task to be immediately
```
"asyncio task" is a slightly confusing term to use here to me. It's appropriate for the `gather_dep`, but I associate executing tasks with threads, not asyncio. I know there's an `execute` coroutine running, but we _could_ easily `.cancel()` that if we wanted to—the limitation is that it can't cancel its thread.

Maybe plain "task" instead of "asyncio task"? I know that's a bit overloaded with user tasks and `TaskState`s as well. "asyncio task or thread"? "underlying operation"?

I understand what you mean right now since I'm familiar with the current implementation and the existence of `execute`, just wondering if someone unfamiliar with the code would be confused why we were only talking about asyncio.